### ISSUE TITLE
--deploy-hook not documented in certbot certonly docs

### ISSUE THREAD

**Issue by Daniel15** [certbot/certbot#6180]
`certbot --help certonly` does not explain that the `--deploy-hook` function can be used to run a deploy hook after issuing the certificate, and that it **also** configures the hook to run on renewals too (by adding it to `/etc/letsencrypt/renewal/example.com.conf` as a `renew_hook` option). This is very useful to know.

**Comment by joohoi** [comment]
Thanks for opening the issue! 

You are right, and we actually have even more hook functionality baked in. There's a longish description of them in [certbot documentation](https://certbot.eff.org/docs/using.html#renewing-certificates). I believe the preferred way of handling the hooks is the hook directories in `/etc/letsencrypt/renewal-hooks` to give better visibility for users what's running and when. I'll bring your point up for discussion with the team.

**Comment by joohoi** [comment]
You can check the certificate within the deploy hook script by using the environment variables `RENEWED_DOMAINS` or `RENEWED_LINEAGE` populated by Certbot to single out the certificate you want to act upon.

**Comment by GrahamH68** [comment]
Are scripts in the deploy directory meant to be called on first creation of a certificate when using CERTONLY with DNS01 ?

I have a simple script to echo out the contents of RENEWED_LINEAGE and RENEWED_DOMAINS, but it doesn't see to be called when a certificate is created.

**Comment by mohag** [comment]
To me it feels as if it might make to not add the `--deploy-hook` value to the config file (especially in the undocumented state). Keeping the `--renew-hook` option (at least on `certonly` and `run` seems like it might result in less unexpected behaviour) (I want to use `--deploy-hook` for a script that installs the certificate initially and it might not be safe to run multiple times...) (e..g append certificate options with filenames at the end of a configuration file)

(If not this, some other way of separating installation and renewal hooks does seem useful for these reasons...)

**Comment by mssaisandeep** [comment]
I am facing the same issue, Running deploy hook for the cert generation will be called sometimes and may not be called sometimes. So is there any hook that I can use for every successful certificate generation(generating/renewal)?? 

BTW if I use certbot certonly --deploy-hook, then that --deploy-hook is getting converted to --renewal-hook in /etc/letsencrypt/renewal/<domainaname>.conf file. 

Please help me.

**Comment by lonix1** [comment]
Has anyone found the answer?

I also want to have a hook run after *cert creation*, but they only reliably run after *cert renewal*.

**Comment by bmw** [comment]
To help answer some questions here and provide more information to help in writing better docs here:

> Are scripts in the deploy directory meant to be called on first creation of a certificate when using CERTONLY with DNS01 ?

@GrahamH68, no. It's kind of buried but from the section https://certbot.eff.org/docs/using.html#renewing-certificates:

> You can also specify hooks by placing files in subdirectories of Certbot’s configuration directory. Assuming your configuration directory is `/etc/letsencrypt`, any executable files found in `/etc/letsencrypt/renewal-hooks/pre`, `/etc/letsencrypt/renewal-hooks/deploy`, and `/etc/letsencrypt/renewal-hooks/post` will be run as pre, deploy, and post hooks respectively when any certificate is **renewed with the renew subcommand**. These hooks are run in alphabetical order and **are not run for other subcommands**.

> (I want to use --deploy-hook for a script that installs the certificate initially and it might not be safe to run multiple times...) (e..g append certificate options with filenames at the end of a configuration file) (On renewal a less drastic script might be run, e..g reload application config)
>
> (If not this, some other way of separating installation and renewal hooks does seem useful for these reasons...)

@mohag, no hooks can be used in this way. You're welcome to open another issue discuss changing this if you like, but I don't think it should be discussed in the same issue as documenting what we currently have.

> I am facing the same issue, Running deploy hook for the cert generation will be called sometimes and may not be called sometimes. So is there any hook that I can use for every successful certificate generation(generating/renewal)??
>
> BTW if I use certbot certonly --deploy-hook, then that --deploy-hook is getting converted to --renewal-hook in /etc/letsencrypt/renewal/.conf file.

@mssaisandeep, you can use `--deploy-hook` to accomplish what you want.

`--deploy-hook` specified on the command line or in a [global configuration file](https://certbot.eff.org/docs/using.html#configuration-file) will be run after obtaining a certificate with any subcommand (e.g. `certonly`, `run`, `renew`).

After a certificate is obtained with a deploy hook specified, it will be written into the renewal configuration file (in `/etc/letsencrypt/renewal` by default). The fact that the `--deploy-hook` is written as a `--renew-hook` is an implementation detail that you can ignore. Once the value has been written to the file like this, you can run `certbot renew` and the same deploy hook be will be run after that certificate is renewed without you having to specify it again. If you have a `--deploy-hook` specified on the command line and a deploy/renew hook in your renewal configuration file, the `--deploy-hook` on the command line is run instead of the command in your renewal configuration file. The value in the renewal configuration file will be changed to the new hook given on the command line once a certificate has successfully been issued.

> Can someone from the team please confirm whether this is a reasonable approach that will not be broken by future changes?

@lonix1, you may also find my previous response useful, but the approach you have taken here will cause that script to be run whenever a certificate is issued with any subcommand. If that's the behavior you want, you're good to go!

**Comment by mohag** [comment]
Documentation page still does not list it under certonly

**Comment by stokito** [comment]
@lonix1 > NOTE! This hook must not also exist in /etc/letsencrypt/renewal-hooks/deploy/ because then it would run twice.

I just checked and it looks like the hook didn't executed twice.
Changed `renew_hook = /etc/letsencrypt/renewal-hooks/deploy/deploy-cert.sh` and moved the hook into the folder. Then made a force renew but it was executed only once.
Maybe this was fixed in code.

**Comment by mohag** [comment]
The parameter is only documented under renew in the man page, with no mention of the hook being added to the config and used on subsequent runs (not sure if that is the case for renew, but it is for certonly)

With a bot closing issues, I'm assuming that +1 comments are now encouraged, since that seems to be the only way to keep the bots at bay....

**Comment by luckman212** [comment]
I'm confused, sorry. I want my `--deploy-hook` script to execute ONCE only upon a NEW cert being obtained. Is method A, B, or C below correct?

### Method A

1) Place deploy script outside of LE dir, e.g. at `/root/my_script.sh`
2) Execute e.g.
```
certbot certonly --force-renewal {...rest of certbot options...} --deploy-hook /root/my_script.sh
```

### Method B

1) Place deploy script _inside_ LE dir, e.g. at `/etc/letsencrypt/renewal-hooks/deploy/my_script.sh`
2) Execute
```
certbot certonly --force-renewal {...rest of certbot options...}
```

### Method C (combines A+B)

1) Place deploy script at `/etc/letsencrypt/renewal-hooks/deploy/my_script.sh`
2) Execute
```
certbot certonly --force-renewal {...rest of certbot options...} --deploy-hook /root/my_script.sh
```

Or something else ???

**Comment by mohag** [comment]
> I'm confused, sorry. I want my `--deploy-hook` script to execute ONCE only upon a NEW cert being obtained. Is method A, B, or C below correct?
> 
> ### Method A
> 1. Place deploy script outside of LE dir, e.g. at `/root/my_script.sh`
> 2. Execute e.g.
> 
> Or something else ???

That and then remove it from he config file for the cert - the renewal hooks specified when the cert is requested is added to the config file...

(I do this by ensuring that this is in the script given for the `--renewal-hook`
```
sed -i '/renew_hook/d' "/etc/letsencrypt/renewal/$(basename "$RENEWED_LINEAGE").conf"
```
)

**Comment by mohag** [comment]
> Ok thanks @mohag
> 
> Just to clarify, even when specifying the `--deploy-hook` commandline argument, the resulting config file ends up containing a line that reads `renew_hook = ...` ?

Correct. (The hooks in the directories that you mentioned gets run without being added to the config file, but they automatically run each renewal) (and very little of this is documented)

**Comment by luckman212** [comment]
Thanks again. For future reference, I found [this post][1] from the LE/certbot forums to be very clear and helpful as well.

[1]: https://community.letsencrypt.org/t/renewal-hooks-per-domain/175621/3

**Comment by amishmm** [comment]
The parameter is only documented under renew in the man page, with no mention of the hook being added to the config and used on subsequent runs (not sure if that is the case for renew, but it is for certonly)

i.e. documentation page still does not list it under certonly

**Comment by jtlarson** [comment]
This issue is still present in certbot 5.0.0 on Ubuntu 24.04.3:

The '--deploy-hook' function is not mentioned in the help commands for 'certonly'

The '--deploy-hook' function listing in the 'renew:' section does not include the fact that it may have been added via inclusion in the initial cert creation command.
